Wordsmith security considerations section (#29)
richsalz/draft-ietf-uta-rfc6125bis: draft-ietf-uta-rfc6125bis.md

## Wildcard Certificates {#security-wildcards}

Wildcard certificates, those that have an identifier with "\*" as the
left-most DNS label, automatically vouch for any single-label host names
within their domain, but not multiple levels of domains.  This can be
convenient for administrators but also poses the risk of vouching for rogue
or buggy hosts. See for example {{Defeating-SSL}} (beginning at slide 91) and
{{HTTPSbytes}} (slides 38-40).

Protection against a wildcard that identifies a public suffix
{{Public-Suffix}}, such as `*.co.uk` or `*.com`, is beyond the scope of this
document.

## Internationalized Domain Names {#security-idn}

Allowing internationalized domain names can lead to visually similar
characters, also referred to as "confusables", being included within
certificates. For discussion, see for example {{IDNA-DEFS, Section 4.4}}
and {{UTS-39}}.

> ## Multiple Identifiers {#security-multi}

Reviewer: ```suggestion
## Multiple Presented Identifiers {#security-multi}
```

A given application service might be addressed by multiple DNS domain names
for a variety of reasons, and a given deployment might service multiple
domains or protocols.  One way to address this issue is for the service
to require the client to send the TLS Server Name Identification
(SNI) extension as discussed in {{TLS, Section 4.4.2.2}}.

> To accommodate the workaround that was needed before the development

Reviewer: Note: GitHub won't allow suggestion to unmodified text, so this is roughly replacing the entire paragraph.

```markdown
This specification allows multiple DNS-IDs, SRV-IDs, or URI-IDs in a
certificate, which allows for a single application service to use the
same certificate for multiple identifiers. This enables a single certificate
to be used across multiple hostnames, such as when a client does not
support the TLS SNI extension, or across multiple protocols, such as
SMTP and HTTP, on a single hostname. The use of a single certificate
and public/private keypair in such environments MAY facilitate
cross-protocol attacks, particularly when used in differing TLS
configurations. See, for example, {{ALPCA}}, {{DROWN}}. Server
operators SHOULD take steps to mitigate the risk of cross-protocol
attacks, such as ensuring all TLS endpoints using a given certificate
support exactly the same TLS version(s) and ciphersuite(s), and
SHOULD make use of the TLS ALPN extension to ensure the correct
protocol is used.

## Multiple Reference Identifiers

This specification describes how a client may construct multiple
acceptable reference identifiers, and may match any of those reference
identifiers with the set of presented identifiers. {{PKIX, Section 4.2.1.10}}
describes a mechanism to allow CA certificates to be constrained in the
set of presented identifiers that they may include within server certificates.
However, these constraints only apply to the explicitly enumerated name
forms. For example, a CA that is only name constrained for DNS-IDs is not
constrained for SRV-IDs and URI-IDs, unless those name forms are also
explicitly included within the name constraints extension.

A client that constructs multiple reference identifiers of different types,
such as both DNS-ID and SRV-IDs, as described in
{#verify-reference-rules}, SHOULD take care to ensure that CAs issuing
such certificates are appropriately constrained. This MAY take the form of
local policy through agreement with the issuing CA, or MAY be enforced by
the client requiring that if one form of presented identifier is constrained,
such as a dNSName name constraint for DNS-IDs, then all other forms of
acceptable reference identities are also constrained, such as requiring a
uniformResourceIndicator name constraint for URI-IDs.
```

This is a huge mouthful, and is really meant to be a rough draft for starting discussion, because it definitely reads awkwardly. However, it tries to get the concept out, but I'm not entirely sure on the usage of the 2119 language here either.

## Wildcard Certificates {#security-wildcards}

Wildcard certificates, those that have an identifier with "\*" as the
left-most DNS label, automatically vouch for any single-label host names
within their domain, but not multiple levels of domains.  This can be
convenient for administrators but also poses the risk of vouching for rogue
or buggy hosts. See for example {{Defeating-SSL}} (beginning at slide 91) and
{{HTTPSbytes}} (slides 38-40).

Protection against a wildcard that identifies a public suffix
{{Public-Suffix}}, such as `*.co.uk` or `*.com`, is beyond the scope of this
document.

## Internationalized Domain Names {#security-idn}

> Allowing internationalized domain names can lead to the inclusion of visually
> similar, or confusable, characters in certificates.  For discussion, see for
> example {{IDNA-DEFS}}.

Reviewer: This is a bit of a nit, but since the reference is to IDNA-DEFS, perhaps it makes sense to keep the sequence ("visually similar characters") together, since that's the title of Section 4.4, since we're not referencing UTR39 directly, which calls them Confusable

```suggestion
Allowing internationalized domain names can lead to visually similar characters, also referred to as "confusables", being included within certificates. For discussion, see for example {{IDNA-DEFS, Section 4.4}}
and {{UTS-39}}.
```

UTS-39 is the primary reference for this these days, and it seems like it's been acceptable to ref informative (e.g. [RFC 5894](https://datatracker.ietf.org/doc/html/rfc5894#ref-Unicode-UTS39)) or normative (e.g. [RFC 8266](https://datatracker.ietf.org/doc/html/rfc8266#ref-UTS39))

Author: Thanks!  I took this wording, and added uts-39 as informative given the text above.

## Multiple Identifiers {#security-multi}

A given application service might be addressed by multiple DNS domain names
for a variety of reasons, and a given deployment might service multiple
domains or protocols.  The client MUST use the TLS Server Name Identification

> (SNI) extension as discussed in {{TLS, Section 4.4.2.2}}.  If multiple
> protocols share the same port, the client MUST use the Application-Layer
> Protocol Negotiation as described in {{ALPN}}.

Reviewer: ```suggestion
(SNI) extension as discussed in {{TLS, Section 4.4.2.2}}.  If multiple
protocols share the same port, the client MUST use the Application-Layer
Protocol Negotiation as described in {{ALPN}} if the intended protocol has
registered Protocol ID {{SOME-REF-HERE}}.
```

Where {{SOME-REF-HERE}} would be reference to https://www.iana.org/assignments/tls-extensiontype-values/tls-extensiontype-values.xhtml#alpn-protocol-ids

There's probably a better way to smith this, and as (about to be) discussed on the list, this could be seen as superfluous.

Author: Yeah, I dropped ALPN completely.
